fhirclient/_utils.py

```diff
@@ -0,0 +1,133 @@
+import urllib
+from typing import Optional, Iterable
+
+import requests
+
+from typing import TYPE_CHECKING, Iterable
+
+if TYPE_CHECKING:
+    from fhirclient.models.bundle import Bundle
+
+
+# Use forward references to avoid circular imports
+def _fetch_next_page(bundle: 'Bundle') -> Optional['Bundle']:
+    """
+    Fetch the next page of results using the `next` link provided in the bundle.
+
+    Args:
+        bundle (Bundle): The FHIR Bundle containing the `next` link.
+
+    Returns:
+        Optional[Bundle]: The next page of results as a FHIR Bundle, or None if no "next" link is found.
+    """
+    next_link = _get_next_link(bundle)
+    if next_link:
+        sanitized_next_link = _sanitize_next_link(next_link)
+        return _execute_pagination_request(sanitized_next_link)
+    return None
+
+
+def _get_next_link(bundle: 'Bundle') -> Optional[str]:
+    """
+    Extract the `next` link from the Bundle's links.
+
+    Args:
+        bundle (Bundle): The FHIR Bundle containing pagination links.
+
+    Returns:
+        Optional[str]: The URL of the next page if available, None otherwise.
+    """
+    if not bundle.link:
+        return None
+
+    for link in bundle.link:
+        if link.relation == "next":
+            return link.url
+    return None
+
+
+def _sanitize_next_link(next_link: str) -> str:
+    """
+    Sanitize the `next` link to ensure it is safe to use.
```

Comment:
I've been testing _sanitize_next_link() more today and noticed a few potential issues. The initial idea behind sanitizing links was to implement a safeguard to protect against URL injection attacks by ensuring the link is safe to use. However, to maintain the library's flexibility, I can't simply rely on blacklisting or whitelisting parameters.

The current approach is just the first step toward safely parsing URLs, but it isn't sufficient for full validation. Users will still need to apply additional validation measures on their end.

On another note, do you think that _sanitize_next_link() is too rigid for an open-source library? 
Am I overthinking it?

Reply:
It might be unnecessary here, especially if it requires some extra client side work to be fully valid.

What are the risks / attack vectors to consider here? The EHR vendor you are connecting to would have to be malicious, yeah? (Like, this URL isn't coming from a human entering data at the hospital - the EHR software directly generates it as part of its search code.)

In which case, the malicious EHR doesn't need to send you to a separate server or mess with URLs. They already have you talking securely to them, and they can leak your credentials or feed you bogus data or whatever it is they want.

Reply:
(I'm assuming that one would be connecting over https and other man-in-the-middle attacks are not in scope?)

Reply:
If man-in-the-middle attacks are in scope - maybe it would be enough to simply validate that the hostnames (netlocs) match between origin_server and the new URL.